src/gridfs/upload.rs

```diff
@@ -173,3 +181,318 @@ impl GridFsBucket {
+// leftover bytes that didn't fill an entire chunk.
+async fn write_bytes(
+    bucket: GridFsBucket,
+    mut buffer: Vec<u8>,
+    starting_n: u32,
+    chunk_size: u32,
+    files_id: Bson,
+) -> Result<(u32, Vec<u8>)> {
+    bucket.create_indexes().await?;
+
+    let mut n = 0;
+    let mut chunks = vec![];
+
+    while buffer.len() as u32 - (n * chunk_size) >= chunk_size {
+        let start = n * chunk_size;
+        let end = (n + 1) * chunk_size;
+        let chunk = Chunk {
+            id: ObjectId::new(),
+            files_id: files_id.clone(),
+            n: starting_n + n,
+            data: RawBinaryRef {
+                subtype: BinarySubtype::Generic,
+                bytes: &buffer[(start as usize)..(end as usize)],
+            },
+        };
+        n += 1;
+        chunks.push(chunk);
+    }
+
+    match bucket.chunks().insert_many(chunks, None).await {
+        Ok(_) => {
+            buffer.drain(..(n * chunk_size) as usize);
+            Ok((n, buffer))
+        }
+        Err(error) => match clean_up_chunks(files_id, bucket.chunks().clone(), Some(error)).await {
+            // clean_up_chunks will always return an error if one is passed in, so this case is
+            // unreachable
+            Ok(()) => unreachable!(),
+            Err(error) => Err(error),
+        },
+    }
+}
+
+async fn close(bucket: GridFsBucket, buffer: Vec<u8>, file: FilesCollectionDocument) -> Result<()> {
+    let insert_result: Result<()> = async {
+        if !buffer.is_empty() {
+            debug_assert!(buffer.len() < file.chunk_size as usize);
+            let final_chunk = Chunk {
+                id: ObjectId::new(),
+                n: file.n() - 1,
+                files_id: file.id.clone(),
+                data: RawBinaryRef {
+                    subtype: BinarySubtype::Generic,
+                    bytes: &buffer[..],
+                },
+            };
+            bucket.chunks().insert_one(final_chunk, None).await?;
+        }
+        bucket.files().insert_one(&file, None).await?;
+        Ok(())
+    }
+    .await;
+
+    match insert_result {
+        Ok(()) => Ok(()),
+        Err(error) => clean_up_chunks(file.id.clone(), bucket.chunks().clone(), Some(error)).await,
+    }
+}
+
+async fn clean_up_chunks(
+    id: Bson,
+    chunks: Collection<Chunk<'static>>,
+    original_error: Option<Error>,
+) -> Result<()> {
+    match chunks.delete_many(doc! { "files_id": id }, None).await {
+        Ok(_) => match original_error {
+            Some(error) => Err(error),
+            None => Ok(()),
+        },
+        Err(error) => {
+            let mut message = String::new();
+            if let Some(original_error) = original_error {
+                message.push_str(&format!("failed operation: {}, then ", original_error));
+            };
+            message.push_str("failed to delete orphaned chunks: {}");
+            Err(ErrorKind::GridFs(GridFsErrorKind::AbortError { message }).into())
+        }
```

Comment:
Sorry for the back-and-forth on this, but it seems like it might be nice to have both of these as `Error` in the `AbortError` variant, and then have the `Display` implementation join them into a string like this. WDYT?

Reply:
updated. I think the naming/docs make it clear enough what's going on but I can also add a custom `Display` impl if you think that would be better